### DISCUSSION TITLE
Limiting network bandwidth at namespace level in RKE2 using Calico

### DISCUSSION THREAD

**Discussion by praseed44** [projectcalico/calico#8962] [General]
Can Calico CNI limit the network bandwidth used by containerd to download images from image registry? In a multi-tenant Kubernetes cluster with Calico as CNI, we are looking for a capability to limit the network bandwidth per namespace to ensure fair usage by all tenants in a cluster. 

Given that, there are huge sized images for ML models running in Kubernetes cluster - image pull of few namespaces is consuming most of the network bandwidth thereby affecting the other tenants in the cluster. 

Looking forward to the feedback from this expert team. Thank you!

**Comment by frozenprocess** [comment]
Calico supports `Traffic shaping plugin`, and you can use it to limit bandwidth via annotations. (More about that [here](https://docs.tigera.io/calico/latest/reference/configure-cni-plugins#cni-network-configuration-lists))
However, the thing that you want to achieve is a bit more complicated given that containerd is shared between all your tenants.

I've got an idea that I'm not sure if it can work but try hostendpoints and traffic shaping maybe that lead to something?
CC: @fasaxc
